### ISSUE TITLE
Fatal concurrency error while starting Nakama (with lua runtimes) under heavy load

### ISSUE THREAD

**Issue by brianwinterpixel** [heroiclabs/nakama#860]
## Description

While a Nakama instance was restarting under heavy load, we encountered this fatal error:

```go
fatal error: concurrent map writes
fatal error: concurrent map writes

goroutine 65087 [running]:
runtime.throw({0x1a541bc, 0xaaaaaaaaaaaaaa})
        runtime/panic.go:1198 +0x71 fp=0xc04e09f1c8 sp=0xc04e09f198 pc=0x992e91
runtime.mapassign_faststr(0x188e960, 0xc0004d4d20, {0xc000a940a0, 0x1e})
        runtime/map_faststr.go:294 +0x38b fp=0xc04e09f230 sp=0xc04e09f1c8 pc=0x97040b
github.com/heroiclabs/nakama/v3/server.newRuntimeLuaVM.func1(0xc000a940a0, {0xc000a940a0, 0x16}, 0xc04e26edc0)
        github.com/heroiclabs/nakama/v3/server/runtime_lua.go:2077 +0xf2 fp=0xc04e09f260 sp=0xc04e09f230 pc=0x16b4c92
github.com/heroiclabs/nakama/v3/server.(*RuntimeLuaNakamaModule).registerRPC(0xc0001e1320, 0x30)
        github.com/heroiclabs/nakama/v3/server/runtime_lua_nakama.go:297 +0x84 fp=0xc04e09f2b0 sp=0xc04e09f260 pc=0x16d02e4
github.com/heroiclabs/nakama/v3/server.(*RuntimeLuaNakamaModule).registerRPC-fm(0xc004446240)
        github.com/heroiclabs/nakama/v3/server/runtime_lua_nakama.go:285 +0x2c fp=0xc04e09f2d0 sp=0xc04e09f2b0 pc=0x17571ec
github.com/heroiclabs/nakama/v3/internal/gopher-lua.callGFunction(0xc004400370, 0x0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:202 +0x37 fp=0xc04e09f338 sp=0xc04e09f2d0 pc=0x1416037
```

<details><summary>Full trace:</summary>

```go
fatal error: concurrent map writes
fatal error: concurrent map writes

goroutine 65087 [running]:
runtime.throw({0x1a541bc, 0xaaaaaaaaaaaaaa})
        runtime/panic.go:1198 +0x71 fp=0xc04e09f1c8 sp=0xc04e09f198 pc=0x992e91
runtime.mapassign_faststr(0x188e960, 0xc0004d4d20, {0xc000a940a0, 0x1e})
        runtime/map_faststr.go:294 +0x38b fp=0xc04e09f230 sp=0xc04e09f1c8 pc=0x97040b
github.com/heroiclabs/nakama/v3/server.newRuntimeLuaVM.func1(0xc000a940a0, {0xc000a940a0, 0x16}, 0xc04e26edc0)
        github.com/heroiclabs/nakama/v3/server/runtime_lua.go:2077 +0xf2 fp=0xc04e09f260 sp=0xc04e09f230 pc=0x16b4c92
github.com/heroiclabs/nakama/v3/server.(*RuntimeLuaNakamaModule).registerRPC(0xc0001e1320, 0x30)
        github.com/heroiclabs/nakama/v3/server/runtime_lua_nakama.go:297 +0x84 fp=0xc04e09f2b0 sp=0xc04e09f260 pc=0x16d02e4
github.com/heroiclabs/nakama/v3/server.(*RuntimeLuaNakamaModule).registerRPC-fm(0xc004446240)
        github.com/heroiclabs/nakama/v3/server/runtime_lua_nakama.go:285 +0x2c fp=0xc04e09f2d0 sp=0xc04e09f2b0 pc=0x17571ec
github.com/heroiclabs/nakama/v3/internal/gopher-lua.callGFunction(0xc004400370, 0x0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:202 +0x37 fp=0xc04e09f338 sp=0xc04e09f2d0 pc=0x1416037
github.com/heroiclabs/nakama/v3/internal/gopher-lua.init.3.func26(0xc004400370, 0x9bab40, 0xc0044231e0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:821 +0x35f fp=0xc04e09f4a8 sp=0xc04e09f338 pc=0x141b05f
github.com/heroiclabs/nakama/v3/internal/gopher-lua.mainLoopWithContext(0xc004400370, 0xc000aa24c0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:60 +0x113 fp=0xc04e09f500 sp=0xc04e09f4a8 pc=0x1415d73
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).callR(0xc004400370, 0x1, 0x1, 0xc000484900)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1211 +0x1b4 fp=0xc04e09f5f8 sp=0xc04e09f500 pc=0x1403bf4
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).Call(...)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1971
github.com/heroiclabs/nakama/v3/internal/gopher-lua.loRequire(0xc004400370)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/baselib.go:559 +0x6b5 fp=0xc04e09f728 sp=0xc04e09f5f8 pc=0x13deed5
github.com/heroiclabs/nakama/v3/internal/gopher-lua.callGFunction(0xc004400370, 0x0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:202 +0x37 fp=0xc04e09f790 sp=0xc04e09f728 pc=0x1416037
github.com/heroiclabs/nakama/v3/internal/gopher-lua.init.3.func26(0xc004400370, 0xb5cff0, 0xc004423140)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:821 +0x35f fp=0xc04e09f900 sp=0xc04e09f790 pc=0x141b05f
github.com/heroiclabs/nakama/v3/internal/gopher-lua.mainLoopWithContext(0xc004400370, 0xc000bc25c0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:60 +0x113 fp=0xc04e09f958 sp=0xc04e09f900 pc=0x1415d73
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).callR(0xc004400370, 0x1, 0x1, 0xc0000bcf90)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1211 +0x1b4 fp=0xc04e09fa50 sp=0xc04e09f958 pc=0x1403bf4
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).Call(...)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1971
github.com/heroiclabs/nakama/v3/internal/gopher-lua.loRequire(0xc004400370)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/baselib.go:559 +0x6b5 fp=0xc04e09fb80 sp=0xc04e09fa50 pc=0x13deed5
github.com/heroiclabs/nakama/v3/internal/gopher-lua.callGFunction(0xc004400370, 0x0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:202 +0x37 fp=0xc04e09fbe8 sp=0xc04e09fb80 pc=0x1416037
github.com/heroiclabs/nakama/v3/internal/gopher-lua.init.3.func26(0xc004400370, 0xa51e00, 0xc0044230a0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:821 +0x35f fp=0xc04e09fd58 sp=0xc04e09fbe8 pc=0x141b05f
github.com/heroiclabs/nakama/v3/internal/gopher-lua.mainLoopWithContext(0xc004400370, 0xc00080bb00)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:60 +0x113 fp=0xc04e09fdb0 sp=0xc04e09fd58 pc=0x1415d73
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).callR(0xc004400370, 0x1, 0x1, 0xc0001afcb0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1211 +0x1b4 fp=0xc04e09fea8 sp=0xc04e09fdb0 pc=0x1403bf4
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).Call(...)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1971
github.com/heroiclabs/nakama/v3/internal/gopher-lua.loRequire(0xc004400370)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/baselib.go:559 +0x6b5 fp=0xc04e09ffd8 sp=0xc04e09fea8 pc=0x13deed5
github.com/heroiclabs/nakama/v3/internal/gopher-lua.callGFunction(0xc004400370, 0x0)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:202 +0x37 fp=0xc04e0a0040 sp=0xc04e09ffd8 pc=0x1416037
github.com/heroiclabs/nakama/v3/internal/gopher-lua.init.3.func26(0xc004400370, 0x9bb860, 0xc004423000)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:821 +0x35f fp=0xc04e0a01b0 sp=0xc04e0a0040 pc=0x141b05f
github.com/heroiclabs/nakama/v3/internal/gopher-lua.mainLoopWithContext(0xc004400370, 0xc01d2cf140)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/vm.go:60 +0x113 fp=0xc04e0a0208 sp=0xc04e0a01b0 pc=0x1415d73
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).callR(0xc004400370, 0x2, 0xffffffffffffffff, 0xc04e0a0398)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1211 +0x1b4 fp=0xc04e0a0300 sp=0xc04e0a0208 pc=0x1403bf4
github.com/heroiclabs/nakama/v3/internal/gopher-lua.(*LState).Call(...)
        github.com/heroiclabs/nakama/v3/internal/gopher-lua/state.go:1971
```

</details>

We are going to try setting min runtimes = max runtimes to avoid this error.

## Steps to Reproduce

1. Configure Nakama with min Lua runtimes = 64, max = 256
2. Start a _lot_ of test clients trying to authenticate repeatedly
3. Start Nakama

## Expected Result

No error. Lua runtimes spin up with no issue.

## Actual Result

Fatal error. See above.

## Context

Godot client

## Your Environment

Nakama 3.10 (forked from https://github.com/heroiclabs/nakama/commit/1d0527e00cdcb5106249e48baa40b03d37bf8b4f)
Postgres
Docker

**Comment by novabyte** [comment]
@brianwinterpixel This is a difficult error for us to help you with because of a couple of challenges:

- You use a forked build of Nakama where we have no insight into the changes you've made.
- You mention "a lot of test clients trying to authenticate repeatedly" but authentication in Nakama does not require Lua server framework execution at all. I think you must have some additional logic attached to that execution flow.

You'll need to share a lot more information for us to be able to find the root cause of the concurrent map writes.

**Comment by brianwinterpixel** [comment]
Yup, my bad. Here are the details:

Our changes to Nakama are limited to:
"add rest interface with default raw json content type" https://github.com/heroiclabs/nakama/pull/775
"Switch order of http_key and Authorization header auth methods" https://github.com/heroiclabs/nakama/pull/772

The clients are attempting to authenticate via AuthenticateCustom. We hook Lua functions into both the before and after hook. After authenticating, the clients then make a couple RPC calls that we have written in Lua as well.
```lua
nk.register_req_before(winterpixel_authenticate.before_authenticate, "AuthenticateCustom")
nk.register_req_after(authenticate_custom, "AuthenticateCustom")
```
```lua
nk.register_rpc(rpc_winterpixel_get_config, "winterpixel_get_config")
```
These functions all do storage operations.

The issue appears to be caused by calls to `nk.register_rpc`, following the stack trace. In Lua we register 76 RPCs and 10 before/after hooks. I (very roughly) estimate the clients were making 1000-2000 requests per second, immediately requiring lua runtimes to allocate to the max configured value.

In our version of Nakama, the stack trace points to runtime_lua.go:2077. This corresponds to this line on the master branch:
https://github.com/heroiclabs/nakama/blob/9b5caf3818c73320f17d94ef5bc0b0b5f2a20e0f/server/runtime_lua.go#L2092

Let me know if you need any more information.

**Comment by zyro** [comment]
@brianwinterpixel Does your Lua code invoke `nk.register_rpc` anywhere inside functions, or is it all top-level within the module? Can you share Lua code that might reproduce this issue at high concurrency?

**Comment by brianwinterpixel** [comment]
Perhaps worth noting, some of our scripts do storage work at the top level.
For example, we read a config.json and write it to storage all guarded by `nk.run_once`. We also read the config for all of our Steam products and write them to storage, except in this case it isn't guarded by `run_once`.

**Comment by brianwinterpixel** [comment]
We believe we've discovered the issue:

```lua
local function rpc_winterpixel_get_config(context, payload)
	local winterpixel_seasons = require("winterpixel.seasons")
	local winterpixel_utils = require("winterpixel.utils")

	local config = query_config()
	config["server_time"] = math.floor(nk.time() / 1000)
	
	local season_data = winterpixel_seasons.compute_season_info(config) 
	winterpixel_utils.table_merge(config, season_data)
	return nk.json_encode(config)
end

nk.register_rpc(rpc_winterpixel_get_config, "winterpixel_get_config")
```

The issue is the `require` call inside the function. This causes the other lua files to load, which themselves re-register some RPCs.

**Comment by zyro** [comment]
That makes sense, if all RPC registrations are at the top level I would not expect any issues. However this pattern would result in RPC registrations at arbitrary execution points.

We'll see if there's potential improvements we can make, until then it would be ideal if you can ensure all registrations are done at the top level. 👍

**Comment by zyro** [comment]
Fixed in 2f99d5a7e70a0d4e0eed6bfd11129b4999057b8d and will be part of the next release.
